### PR TITLE
Revise natives code

### PR DESCRIPTION
**Pull request by Kenzzer**
# What's changed ?

Instead of all the natives being inside various include files, and then all included by `natives.h` we instead divide those natives into multiple `.hpp` and `.cpp` files, available in the `natives` directory. Very much mirroring the setup of our .inc files.

# New way to register natives

No more global array, instead an std::vector is passed around by reference. The natives are then added to it by each `.cpp` module. Furthermore, macros to define native functions have been eliminated. Everything are now locally defined functions, only available in their respective `.cpp` file.

Example :
```cpp
void setup(std::vector<sp_nativeinfo_t>& natives) {
	area::setup(natives);
	surroundingareas::setup(natives);
	
	tf::nav::setup(natives);

	sp_nativeinfo_t list[] = {
		{"CNavMesh.Address.get", GetAddress},
		{"CNavMesh.IsLoaded", IsLoaded},
		{"CNavMesh.IsAnalyzed", IsAnalyzed},
		{"CNavMesh.IsOutOfDate", IsOutOfDate},
		{"CNavMesh.GetNavAreaCount", GetNavAreaCount},
		{"CNavMesh.CollectSurroundingAreas", CollectSurroundingAreas},
		{"CNavMesh.GetNavAreaByID", GetNavAreaByID},
		{"CNavMesh.GetNearestNavArea", GetNearestNavArea},
		{"CNavMesh.BuildPath", BuildPath},
		{"CNavMesh.GetNavArea", GetNavArea},
		{"CNavMesh.GetNavAreaEntity", GetNavAreaEntity},
	};

	natives.insert(natives.end(), std::begin(list), std::end(list));
}
```

# New coding style for `{}`

As of this PR, I wish that we adopt Rust's styling for `.cpp` and `.hpp` files. The extension never had a coding style to begin with, so I believe this is a good point to start introducing some.

# New/updated natives

- Added the following natives to the extension. The usefulness and 0 gamedata cost more than enough warrants the addition of such a useful methodmap to the extension.
```
CTakeDamageInfo.CTakeDamageInfo
GetInflictor
SetInflictor
GetWeapon
SetWeapon
GetAttacker
SetAttacker
GetDamage
SetDamage
GetMaxDamage
SetMaxDamage
ScaleDamage
AddDamage
SubtractDamage
GetDamageBonus
SetDamageBonus
GetDamageBonusProvider
GetBaseDamage
BaseDamageIsValid
GetDamageForce
SetDamageForce
ScaleDamageForce
GetDamageForForceCalc
SetDamageForForceCalc
GetDamagePosition
SetDamagePosition
GetReportedPosition
SetReportedPosition
GetDamageType
SetDamageType
AddDamageType
GetDamageCustom
SetDamageCustom
GetDamageStats
SetDamageStats
SetForceFriendlyFire
IsForceFriendlyFire
GetAmmoType
SetAmmoType
GetAmmoName
GetPlayerPenetrationCount
SetPlayerPenetrationCount
GetDamagedOtherPlayers
SetDamagedOtherPlayers
SetCritType
GetCritType
```

- Fixed `CBaseNPC` and `CExtNPC` natives not verifying invalid npc index. Leading to spectacular nullpointer crashes.

# Prepare for TF2 x64 bits

In addition to everything above, speculative but correct changes have been made to hopefully prepare enough the extension for the eventual release of 64bits tf2. See alliedmodders/hl2sdk#127 for further details.

### CODE REVIEW

File: extension/natives/nextbot/component.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
+#include "path.hpp"
+
+namespace natives::nextbot::component {
+
+inline INextBotComponent* Get(IPluginContext* context, const cell_t param) {	
+	INextBotComponent* component = (INextBotComponent*)param;
```
Comment (KitRifty) [inline extension/natives/nextbot/component.cpp:6]:
```suggestion
	INextBotComponent* component = (INextBotComponent*)PawnAddressToPtr(param);
```

File: extension/natives/nextbot/component.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+	return 0;
+}
+
+cell_t GetBot(IPluginContext* context, const cell_t* params) {
+	auto component = Get(context, params[1]);
+	if (!component) {
+		return 0;
+	}
+
+	return (cell_t)(component->GetBot());
```
Comment (KitRifty) [inline extension/natives/nextbot/component.cpp:51]:
```suggestion
	return PtrToPawnAddress(component->GetBot());
```

File: extension/natives/nextbot/eventresponder.cpp [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+		return context->ThrowNativeError("Entity %d is invalid!", params[2]);
+	}
+
+	auto bcc = entity->MyCombatCharacterPointer();
+	if (!bcc) {
+		return context->ThrowNativeError("Entity %d is not a CBaseCombatCharacter!", params[2]);
+	}
+
+	CBaseEntity* weapon = gamehelpers->ReferenceToEntity(params[3]);
+	if (!weapon && params[2] != -1) {
```
Comment (KitRifty) [inline extension/natives/nextbot/eventresponder.cpp:220]:
```suggestion
	if (!weapon && params[3] != -1) {
```

File: extension/natives/nextbot/eventresponder.cpp [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+
+cell_t OnShoved(IPluginContext* context, const cell_t* params) {
+	auto responder = Get(context, params[1]);
+	if (!responder) {
+		return 0;
+	}
+
+	CBaseEntity* entity = gamehelpers->ReferenceToEntity(params[2]);
+	if (!entity && params[2] != -1)
+	{
```
Comment (KitRifty) [inline extension/natives/nextbot/eventresponder.cpp:407]:
```suggestion
	if (!entity && params[2] != -1) {
```

File: extension/natives/nextbot/intention.cpp [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+	INextBot *me = intention->GetBot();
+	if (!me) {
+		return context->ThrowNativeError("GetBot returned null!");
+	}
+
+	Vector targetPoint = intention->SelectTargetPoint(me, subject);
+	buffer[0] = sp_ftoc( targetPoint[0] );
+	buffer[1] = sp_ftoc( targetPoint[1] );
+	buffer[2] = sp_ftoc( targetPoint[2] );
```
Comment (KitRifty) [inline extension/natives/nextbot/intention.cpp:132]:
```suggestion
	VectorToPawnVector( buffer, targetPoint );
```

File: extension/natives/nextbot/intention.cpp [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	if (!pos) {
+		return context->ThrowNativeError("Position cannot be null !");
+	}
+
+	INextBot *me = intention->GetBot();
+	if (!me) {
+		return context->ThrowNativeError("GetBot returned null!");
+	}
+
+	Vector vecPos(sp_ctof(pos[0]), sp_ctof(pos[1]), sp_ctof(pos[2]));
```
Comment (KitRifty) [inline extension/natives/nextbot/intention.cpp:154]:
```suggestion
	Vector vecPos;
	PawnVectorToVector( pos, vecPos );
```

File: extension/natives/nextbot/locomotion.cpp [hunk 7]
```diff
@@ -0,0 +1,713 @@
...
+	return 0;
+}
+
+cell_t ClimbLadder(IPluginContext* context, const cell_t* params) {
+	auto mover = Get(context, params[1]);
+	if (!mover) {
+		return 0;
+	}
+
+	mover->ClimbLadder((CNavLadder *)(params[2]), (CNavArea *)(params[3]));
```
Comment (KitRifty) [inline extension/natives/nextbot/locomotion.cpp:266]:
```suggestion
	mover->ClimbLadder((CNavLadder *)PawnAddressToPtr(params[2]), (CNavArea *)PawnAddressToPtr(params[3]));
```

File: extension/natives/nextbot/locomotion.cpp [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+	return 0;
+}
+
+cell_t DescendLadder(IPluginContext* context, const cell_t* params) {
+	auto mover = Get(context, params[1]);
+	if (!mover) {
+		return 0;
+	}
+
+	mover->DescendLadder((CNavLadder *)(params[2]), (CNavArea *)(params[3]));
```
Comment (KitRifty) [inline extension/natives/nextbot/locomotion.cpp:276]:
```suggestion
	mover->DescendLadder((CNavLadder *)PawnAddressToPtr(params[2]), (CNavArea *)PawnAddressToPtr(params[3]));
```

File: extension/natives/nextbot/path.cpp [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+cell_t GetPosition(IPluginContext* context, const cell_t* params) {
+	auto path = Get<Path>(context, params[1]);
+	if (!path) {
+		return 0;
+	}
+
+	float dist = sp_ctof(params[2]);
+	cell_t* addr;
+	context->LocalToPhysAddr(params[3], &addr);
+	Path::Segment* segment = (Path::Segment*)(params[4]);
```
Comment (KitRifty) [inline extension/natives/nextbot/path.cpp:299]:
```suggestion
	Path::Segment* segment = (Path::Segment*)PawnAddressToPtr(params[4]);
```

File: extension/natives/nextbot/path.cpp [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+	return 0;
+}
+
+cell_t Copy(IPluginContext* context, const cell_t* params) {
+	auto path = Get<Path>(context, params[1]);
+	if (!path) {
+		return 0;
+	}
+
+	INextBot* bot = (INextBot*)(params[2]);
```
Comment (KitRifty) [inline extension/natives/nextbot/path.cpp:312]:
```suggestion
	INextBot* bot = (INextBot*)PawnAddressToPtr(params[2]);
```

File: extension/natives/nextbot/path.cpp [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+	Path* path2 = (Path*)(params[3]);
```
Comment (KitRifty) [inline extension/natives/nextbot/path.cpp:313]:
```suggestion
	Path* path2 = (Path*)PawnAddressToPtr(params[3]);
```

File: extension/natives/nextbot/path.cpp [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+		return 0;
+	}
+
+	cell_t* nearAddr;
+	context->LocalToPhysAddr(params[2], &nearAddr);
+	Vector vecNear;
+	PawnVectorToVector(nearAddr, vecNear);
+	cell_t* posAddr;
+	context->LocalToPhysAddr(params[3], &posAddr);
+	Path::Segment* segment = (Path::Segment*)(params[4]);
```
Comment (KitRifty) [inline extension/natives/nextbot/path.cpp:331]:
```suggestion
	Path::Segment* segment = (Path::Segment*)PawnAddressToPtr(params[4]);
```

File: extension/natives/nextbot/path.cpp [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+	return 0;
+}
+
+cell_t Draw(IPluginContext* context, const cell_t* params) {
+	auto path = Get<Path>(context, params[1]);
+	if (!path) {
+		return 0;
+	}
+
+	Path::Segment* segment = (Path::Segment*)(params[2]);
```
Comment (KitRifty) [inline extension/natives/nextbot/path.cpp:480]:
```suggestion
	Path::Segment* segment = (Path::Segment*)PawnAddressToPtr(params[2]);
```

**Review by KitRifty — CHANGES_REQUESTED** [review]
Very nice work refactoring the natives code. This is much preferred over having that huge list in the main natives file. However there are a couple of types and a handful of instances where pawn addresses aren't being converted to pointers, and vice versa.

File: extension/natives/takedamageinfo.cpp [hunk 14]
```diff
@@ -0,0 +1,570 @@
...
+		//{"CTakeDamageInfo.GetAmmoName", GetAmmoName},
+
+		{"CTakeDamageInfo.GetPlayerPenetrationCount", GetPlayerPenetrationCount},
+		{"CTakeDamageInfo.SetPlayerPenetrationCount", SetPlayerPenetrationCount},
+
+		{"CTakeDamageInfo.GetDamagedOtherPlayers", GetDamagedOtherPlayers},
+		{"CTakeDamageInfo.SetDamagedOtherPlayers", SetDamagedOtherPlayers},
+
+		{"CTakeDamageInfo.SetCritType", SetCritType},
+		{"CTakeDamageInfo.GetCritType", GetCritType},
```
[the remark this answers is not in the document]
Comment (Kenzzer) [reply extension/natives/takedamageinfo.cpp:564]: The natives still need to be registered otherwise plugins will fail to load.

**Review by KitRifty — APPROVED** [review]

**Pull request merged**